code/iaas/logic/src/main/java/io/cattle/platform/process/host/HostActivatePreListener.java

```diff
@@ -0,0 +1,97 @@
+package io.cattle.platform.process.host;
+
+import io.cattle.iaas.cluster.process.lock.ClusterLock;
+import io.cattle.platform.core.constants.ClusterConstants;
+import io.cattle.platform.core.constants.CommonStatesConstants;
+import io.cattle.platform.core.dao.ClusterHostMapDao;
+import io.cattle.platform.core.model.Host;
+import io.cattle.platform.core.model.tables.records.ClusterHostMapRecord;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.lock.LockCallback;
+import io.cattle.platform.lock.LockManager;
+import io.cattle.platform.object.process.StandardProcess;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+@Named
+public class HostActivatePreListener extends AbstractObjectProcessLogic implements ProcessPreListener {
+
+    @Inject
+    ClusterHostMapDao clusterHostMapDao;
+
+    @Inject
+    LockManager lockManager;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "host.activate" };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        if (!objectManager.isKind(state.getResource(), ClusterConstants.KIND)) {
+            return new HandlerResult(false, (Map<Object, Object>)null);
+        }
+        final Host cluster = (Host)state.getResource();
+
+        return lockManager.lock(new ClusterLock(cluster), new LockCallback<HandlerResult>() {
+            @Override
+            public HandlerResult doWithLock() {
+                // check if this cluster's already managed by a specific host
+                Long hostId = DataAccessor.fields(cluster).withKey(ClusterConstants.MANAGING_HOST).as(Long.class);
+
+                // if not, find suitable host within cluster to manage the cluster server
+                // This actually doesn't check whether the current host is actually in the cluster. Which isn't really a requirement anyways.
+                if (hostId == null || !isHostActive(hostId)) {
+                    hostId = findSuitableHost(cluster);
```

Comment:
Sorry, rephrasing - scheduling deactivation logic (lets call it process #1) can be put to the separate handler invoked on managing host set to Null (process #2). Because process #2 itself can be triggered by:

1) host is deactivated
2) host is removed from cluster
3) host is removed from cattle
4) ... and probably more

Just to make the code cleaner, otherwise you will have to duplicate executing process #1 and scheduling process #2 in 1) 2) 3) and 4)